[Merged by Bors] - Fix bad assumption when checking finalized descendant

paulhauner opened the pull request:
## Issue Addressed

- Resolves #1616

## Proposed Changes

Fixes a bug where we are unable to read the finalized block from fork choice.

## Detail

I had made an assumption that the finalized block always has a parent root of `None`:

https://github.com/sigp/lighthouse/blob/e5fc6bab485fa54d7e518b325f4eb9201ba5c6a1/consensus/fork_choice/src/fork_choice.rs#L749-L752

This was a faulty assumption, we don't set parent *roots* to `None`. Instead we *sometimes* set parent *indices* to `None`, depending if this pruning condition is satisfied: 

https://github.com/sigp/lighthouse/blob/e5fc6bab485fa54d7e518b325f4eb9201ba5c6a1/consensus/proto_array/src/proto_array.rs#L229-L232 

The bug manifested itself like this:

1. We attempt to get the finalized block from fork choice
1. We try to check that the block is descendant of the finalized block (note: they're the same block).
1. We expect the parent root to be `None`, but it's actually the parent root of the finalized root.
1. We therefore end up checking if the parent of the finalized root is a descendant of itself. (note: it's an *ancestor* not a *descendant*).
1. We therefore declare that the finalized block is not a descendant of (or eq to) the finalized block. Bad.

## Additional Info

In reflection, I made a poor assumption in the quest to obtain a probably negligible performance gain. The performance gain wasn't worth the risk and we got burnt.

michaelsproul reviewed (approved):
Looks good. Agree with the fix.

Is there a simple place we could chuck a unit test for this?

paulhauner commented:
> Is there a simple place we could chuck a unit test for this?

Yep, good call. I added a regression test that fails on master but passes on this branch.

I also replaced one of the `ParentUnknown` errors, as you suggested out-of-band yesterday: https://github.com/sigp/lighthouse/pull/1629/commits/0de48efe97a9937dc14407b280e9997232d826ae.

michaelsproul reviewed (approved):
Perfecto, merge at will!

paulhauner commented:
bors r+

paulhauner commented:
bors retry

paulhauner commented:
bors retry

why u so naughty bors?

paulhauner commented:
bors r-

paulhauner commented:
bors r+

michaelsproul commented:
Hmm, maybe we should bump the 2hr Bors timeout, looks like the macOS build took 2hrs 8mins https://github.com/sigp/lighthouse/runs/1131898681?check_suite_focus=true
